### PR TITLE
Api batch endpoint

### PR DESCRIPTION
**Pull request by carlfeberhard**
cross reference: https://trello.com/c/akpENftl/836-api-batch-requests

This is something that would really improve a few sections of the UI to reduce requests. But it's a rough design and I'm open to suggestions for any security/portablility/production implications or problems people see.

### CODE REVIEW

**Comment by dannon** [comment]
@carlfeberhard I saw galaxybot added the 16.07 milestone to this -- I think you had said this would wait until 16.10 right?  Just wanted to doublecheck before I swap it.

**Comment by carlfeberhard** [comment]
@dannon yes - you're right. Is there a reason the bot does this?

**Comment by martenson** [comment]
@carlfeberhard it was misconfigured, I fixed it

**Comment by carlfeberhard** [comment]
@martenson Cool - thanks for that

**Comment by dannon** [comment]
@carlfeberhard I like this approach, and it's working great for me.  Thanks for fielding my PMs on IRC.

**Comment by carlfeberhard** [comment]
@dannon Thanks

File: lib/galaxy/web/framework/middleware/batch.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+        # if we set it, webob will try to use the buffer interface on a unicode string
+        request_environ[ 'CONTENT_LENGTH' ] = None
+        # this may well need to change in py3
+        request_body = io.BytesIO( bytearray( request_body, encoding='utf8' ) )
+        request_environ[ 'wsgi.input' ] = request_body
+        # log.debug( 'request_environ:\n%s', pprint.pformat( request_environ ) )
+
+        return request_environ
+
+    def _proccess_batch_request( self, request, environ, start_response ):
```
Comment (nsoranzo) [inline lib/galaxy/web/framework/middleware/batch.py:152]: s/proccess/process/

File: lib/galaxy/web/framework/middleware/batch.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
+"""
+Batch API middleware
+
+Adds a single route to the installation that:
+  1. accepts a POST call containing a JSON array of 'http-like' JSON
+     dictionaries.
+  2. Each dictionary describes a single API call within the batch and is routed
+     back by the middleware to the application's `handle_request` as if it was
+     a seperate request.
```
Comment (nsoranzo) [inline lib/galaxy/web/framework/middleware/batch.py:9]: s/seperate/separate/

**Comment by nsoranzo** [comment]
@carlfeberhard Cool stuff, thanks!

**Pull request merged**